python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+        # See if we can fit this job into our busy schedule
+        if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
+            return True
+
+        # Check for insufficient slots -soft limit
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and self.soft_limit:
+            tester.specs.addParam('caveats', ['OVERSIZED'], "")
+            return True
+
+        # Check for insufficient slots -hard limit (skip this job)
+        # TODO: Create a unit test for this case
+        elif tester.getProcs(self.options) > self.job_slots and not self.soft_limit:
+            tester.setStatus('insufficient slots', tester.bucket_skip)
+
+    # Status processing (hand a tester's status back to the TestHarness)
+    def statusRunner(self, queue):
+        tester_name = queue.get()
+        tester = self.scheduled_name_to_object[tester_name]
+
+        # Test is Finished (failed, passed, skipped, silent, etc)
+        if tester.isFinished():
+            # And the running caveat
+            if tester_name in self.reported_jobs and tester.didPass():
+                tester.specs.addParam('caveats', ['FINISHED'], "")
+
+            # Print finalized results
+            self.harness.handleTestStatus(tester)
+
+            # Remove this finished tester from the DAG and active job set, get the next node, and assign it to
+            # a runner thread
+            with self.thread_lock:
+                tester_dag = self.shared_testers_dag[tester.getTestDir()]
+
+                # Handle dependencies that will fail if this test did not successfully finish but only if we care
+                # about prereqs
+                if not tester.didPass():
+                    self.deleteDownstreamTests(tester_dag, tester)
+
+                # Get the next list of runnable nodes (jobs)
+                tester_dag.delete_node(tester_name)
+                next_jobs = tester_dag.ind_nodes()
+
+                # Loop through and determine if we need to assign a thread to this new job
+                new_tmp_jobs = []
+                for job in next_jobs:
+                    tmp_tester = self.scheduled_name_to_object[job]
+                    if job not in self.active_jobs and tmp_tester.isInitialized():
+                        tmp_tester.setStatus('QUEUED', tmp_tester.bucket_pending)
+                        new_tmp_jobs.append(job)
+
+            # Add new jobs outside thread lock
+            for job in new_tmp_jobs:
+                self.runnerGo(job)
+
+        # Test is Pending and could not run for some reason or another
+        else:
+            # The runner when performing checkRunnableBase, didn't fail or skip the test. But it couldn't
+            # run for other reasons (not enough slots, etc). So place it back in the runner queue
+            if tester_name not in self.active_jobs:
+                self.runnerGo(tester_name)
```

Comment:
Not crazy about this name, doesn't tell use anything useful. Maybe `insertRunnerQueue`?